### PR TITLE
Synchronized the discovery document retrieving

### LINKED ISSUE #372: Discovery Endpoint caching not working when multiple threads consume GetAsync
Hello!

Related [issue](https://github.com/IdentityModel/IdentityModel/issues/201). This has been closed cause the user didn't provide examples nor deeper explanation, but I ran into the same situation and I was able to isolate the case and understand why.

It happens that the DiscoveryCache is not thread safe, then when the app starts under heavy load, there might be cases where multiple threads would await on GetAsync, resulting on multiple request made. Since it is not thread safe, this causes multiple request to the Discovery Endpoint the first time the document needs to be refreshed or fetched.

After that, it works as expected.

I tested this scenario under 2 different conditions:

IIS 10.0 and .Net Framework 4.8: using Autofac registering the service as Singleton
Docker and .Net Core 3.1: again, registering the service as Singleton
Looking at the code, I see why this is happening:

```
public Task<DiscoveryDocumentResponse> GetAsync()
{
    // If multiple threads get to this point while 
    // the request is being made, a new task per
    // thread will be returned
    if (_nextReload <= DateTime.UtcNow)
    {
        Refresh();
    }

    return _lazyResponse.Value;
}

public void Refresh()
{
    _lazyResponse = new AsyncLazy<DiscoveryDocumentResponse>(GetResponseAsync);
}
```

The current implementation preseted a challenge to me, since it is actually caching the task rather than the document, I first created a wrapper around this cache to cache the result of task. But when I compared that code with the current implementation of the DiscoveryCache, I just realized that it was simpler to create a brand new thread safe caching implementation. 

I feel like I am missing something or I am not understanding how to use the cache properly.

Any suggestions?

I really appreciate your help.
Thank you very much!!

### PR DESCRIPTION
**Pull request by AndreReise**
As it was mentioned in #372  - `DiscoveryCache` is not thread-safe.

To prevent  sending multiple `Get` request to discovery endpoint a simple binary `Semaphore Slim` synchronization was added.

### CODE REVIEW

**Comment by leastprivilege** [comment]
@brockallen could you review for the 6.1 release?

File: src/Client/DiscoveryCache.cs [hunk 1]
```diff
@@ -54,14 +56,22 @@ public DiscoveryCache(string authority, Func<HttpMessageInvoker> httpClientFunc,
     /// Get the DiscoveryResponse either from cache or from discovery endpoint.
     /// </summary>
     /// <returns></returns>
-    public Task<DiscoveryDocumentResponse> GetAsync()
+    public async Task<DiscoveryDocumentResponse> GetAsync()
     {
-        if (_nextReload <= DateTime.UtcNow)
+        await _semaphoreSlim.WaitAsync()
```
Comment (josephdecock) [inline src/Client/DiscoveryCache.cs:61]: We'd like to have a timeout for the semaphore's wait. We'd need to figure out where that would come from though.

File: src/Client/DiscoveryCache.cs [hunk 2]
```diff
@@ -54,14 +56,22 @@ public DiscoveryCache(string authority, Func<HttpMessageInvoker> httpClientFunc,
...
+            .ConfigureAwait(false);
+
+        try
         {
-            Refresh();
-        }
+            var discoveryDocument = await GetAsyncCore()
+                .ConfigureAwait(false);
```
Comment (josephdecock) [inline src/Client/DiscoveryCache.cs:67]: Prefer our helper extension method to ConfigureAwait without the parameter

File: src/Client/DiscoveryCache.cs [hunk 3]
```diff
@@ -72,13 +82,27 @@ public void Refresh()
...
+    }
+
     private async Task<DiscoveryDocumentResponse> GetResponseAsync()
     {
         var result = await _getHttpClient().GetDiscoveryDocumentAsync(new DiscoveryDocumentRequest
         {
             Address = _authority,
             Policy = _policy
-        }).ConfigureAwait();
+        }).ConfigureAwait(false);
```
Comment (josephdecock) [inline src/Client/DiscoveryCache.cs:105]: Prefer our helper extension method to ConfigureAwait without the parameter

**Review by josephdecock — COMMENTED** [review]
@leastprivilege - do you think that we should be trying to add thread safety on the discovery cache at all? The argument against is: simpler code is easier to maintain, I'm scared about deadlocks, and all consumers have to pay the cost of acquiring the lock even if they don't need thread safety.

**Comment by josephdecock** [comment]
> The argument against is: simpler code is easier to maintain, I'm scared about deadlocks, and all consumers have to pay the cost of acquiring the lock even if they don't need thread safety.

I'm closing this PR for these reasons, and also because I believe that if you need a discovery cache guaranteed to only allow a single thread to make the discovery request, it should be possible to wrap the existing discovery cache in this kind of locking outside IdentityModel itself.

We could revisit this issue, especially if wrapping the cache with locks like I mention isn't feasible. If we were to do that, I'd like the locking to be in a new type, separate from the existing DiscoveryCache - a ConcurrentDiscoveryCache;.

**Pull request closed without merging**